Add user guide for cross-account lambda functionality (#8180)
solo-io/gloo: docs/content/guides/traffic_management/destination_types/aws_lambda/cross-account.md

---
title: Using Cross-Account Lambda Functions
weight: 101
description: |
  This guide describes how to configure Gloo Edge to route to AWS Lambda functions in different accounts than the one used to authenticate with AWS.
---

# Overview
 - Gloo Edge supports routing to AWS Lambda functions in different accounts than the one used to authenticate with AWS
 - There are two strategies that can be used to configure this behavior:
   1. [IAM Roles for Service Accounts (IRSA) Configuration](#iam-roles-for-service-accounts-irsa-configuration) (Recommended)
   2. [Resource-Based Configuration](#resource-based-configuration)

# IAM Roles for Service Accounts (IRSA) Configuration
 - This is the recommended workflow for using cross-account Lambda functions with Gloo Edge

>  - The configuration is essentially a modified version of the AWS Lambda with EKS ServiceAccounts [guide](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/aws_lambda/eks-service-accounts/)

Reviewer: versioned link paths
```suggestion
 - The configuration is essentially a modified version of the AWS Lambda with EKS ServiceAccounts [guide]({{< versioned_link_path fromRoot="/guides/traffic_management/destination_types/aws_lambda/eks-service-accounts/" >}})
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

> ## AWS Configuration

Reviewer: Could we add a summary of what you do in this section? maybe something like:
```suggestion
## AWS Configuration
Your primary AWS account has a role with permissions to invoke Lambda functions in one or more target accounts. Then, you add the role and Lambda function details to the Gloo Edge Upstream resource.

```

Reviewer: Adding the diagram you showed in the demo might also be a nice conceptual overview :)

Author: Added clarification around this here: https://github.com/solo-io/gloo/pull/8180/commits/50cc3a744442fe4e582c7fda6d4ab1ae48def5e6

### Primary Account

>   - Follow the steps in the [AWS Lambda with EKS ServiceAccounts guide](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/aws_lambda/eks-service-accounts/)

Reviewer: ```suggestion
  - Follow the steps in the [AWS Lambda with EKS ServiceAccounts guide]({{< versioned_link_path fromRoot="/guides/traffic_management/destination_types/aws_lambda/eks-service-accounts/" >}})
```

>     - As part of this guide, you will create a role which will be associated with the ServiceAccount in your cluster. This role will be used to assume the role in the target account, which will be used to invoke the Lambda function.

Reviewer: ```suggestion
    - As part of this guide, you will create a role which will be associated with the ServiceAccount in your cluster. This role will be used to assume the role in the target account, which will be used to invoke the Lambda function. Make sure to note the `ARN` of this role to use when configuring the target account's role later.
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

### Target Account
 - Create a Lambda function in the target account
 - Create a role which:

>    1. Can be used to invoke the Lambda function -- this role should have the `lambda:InvokeFunction` permission

Reviewer: ```suggestion
   1. Can be used to invoke the Lambda function, which requires the `lambda:InvokeFunction` permission
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

   1. Can be assumed by the role associated with the ServiceAccount in the primary account

>      - in order to do this, specify the following trust policy on the role:

Reviewer: ```suggestion
     - To do so, specify the following trust policy on the role with the `ARN` of the primary account's role that you previously retrieved:
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

       ```json
        {
            "Version": "2012-10-17",
            "Statement": [
                {
                    "Effect": "Allow",
                    "Principal": {
                        "AWS": <ARN OF PRIMARY ACCOUNT ROLE>,
                    },
                    "Action": "sts:AssumeRole",
                    "Condition": {}
                }
            ]
        }
       ```
    
## Gloo Edge Configuration

>  - Disable Function Discovery (FDS). This can be used to automatically discover functions in the user's primary AWS account, but since we are using a secondary account, we will need to manually configure the functions.

Reviewer: ```suggestion
 - Disable Function Discovery (FDS), which automatically discovers functions in your primary AWS account. Because you want to route to Lambdas in more than one account, you must disable discovery and instead manually configure the functions.
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

   - This can be done via the `gloo.discovery.fdsMode` setting to `DISABLED` in the enterprise Helm chart
   - Alternatively, you can set `spec.discovery.fdsMode` to `DISABLED` in the `gloo.solo.io/v1.Settings` custom resource
 - Specify the Lambda functions in the target account that you wish to route to on the upstream
   - These are defined under the `spec.aws.LambdaFunctions` field of the upstream

> - Configure the `spec.aws.roleArn` field of the upstream to point to the IAM role that will be used to invoke the Lambda functions (I.E., the role created in the target account)

Reviewer: ```suggestion
- Configure the `spec.aws.roleArn` field of the upstream to point to the IAM role that will be used to invoke the Lambda functions, which is the role that you created in the target account
```

Author: Suggestion applied here: https://github.com/solo-io/gloo/pull/8180/commits/c72e822c378d6c008c128276db1c37942927d6e4

  - This role should have the `lambda:InvokeFunction` permission
  - Example Upstream:
      ```yaml
      apiVersion: gloo.solo.io/v1
      kind: Upstream
      metadata:
        name: aws-upstream
        namespace: gloo-system
      spec:
        aws:
          region: us-east-1
          secretRef:
            name: aws-creds
            namespace: gloo-system
          roleArn: arn:aws:iam::123456789012:role/lambda-role
          # because we have disabled FDS, we need to hardcode the lambda functions in the upstream spec
          lambdaFunctions:
          - lambdaFunctionName: target-name
            logicalName: target-name
      ```

# Resource-Based Configuration
## AWS Configuration
### Primary Account
 - Create a user or role in the primary account that will be used to invoke the Lambda functions in the secondary account
   - Give the user the `lambda:InvokeFunction` permission

>    - Create an access key for the user

Reviewer: I am wondering what the access key here is for? I don't see it getting used later...is it related to the ARN that gets used in the target account?

Author: Clarified this point here: https://github.com/solo-io/gloo/pull/8180/commits/4e619fecd3cef6bb7f54685e3027b801391093ba

   - This can be done via the `gloo.discovery.fdsMode` setting to `DISABLED` in the enterprise Helm chart
   - Alternatively, you can set `spec.discovery.fdsMode` to `DISABLED` in the `gloo.solo.io/v1.Settings` custom resource
 - Specify the Lambda functions in the target account that you wish to route to on the upstream
   - These are defined under the `spec.aws.LambdaFunctions` field of the upstream 
 - Specify the target account ID in the upstream
   - This is defined under the `spec.aws.awsAccountId` field of the upstream
 - Example Upstream:
     ```yaml
     apiVersion: gloo.solo.io/v1
     kind: Upstream
     metadata:
       name: aws-upstream
       namespace: gloo-system
     spec:
       aws:
         region: us-east-1
         secretRef:
           name: aws-creds
           namespace: gloo-system
         awsAccountId: "123456789012"
         # because we have disabled FDS, we need to hardcode the lambda functions in the upstream spec
         lambdaFunctions:
         - lambdaFunctionName: target-name
           logicalName: target-name

>      ```

Reviewer: Is there anything they can do afterwards to verify that the cross-account lambda routing works?

Maybe it would just be following the rest of this guide? https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/aws_lambda/eks-service-accounts/#routing-to-your-lambdas

Author: Added verification steps here: https://github.com/solo-io/gloo/pull/8180/commits/0ce4f3fc8e9319147f24e9a2c5f9f255af1e51db
